src/metabase/api/dashboard.clj

```diff
@@ -244,7 +247,35 @@
   [dashboard]
   (update dashboard :dashcards add-query-average-duration-to-dashcards))
 
-(defn- get-dashboard
+;; ## Dashboard load caching
+;; When the FE loads a dashboard, there is a burst of requests sent to the BE:
+;; - One  /api/dashboard/:id
+;; - One  /api/dashboard/:id/query_metadata
+;; - Many /:dashboard-id/dashcard/:dashcard-id/card/:card-id/query
+;; Each of these needs some metadata from the appdb: to hydrate the dashboard, get the query_metadata, and to run the
+;; query processor over all the dashcards.
+;; That leads to a lot of re-fetching of the same information from the appdb, and is a great opportunity for caching.
+;;
+;; To connect the dots across these N+2 HTTP requests, the FE attaches a `?dashboard_load_id=UUID` parameter to all
+;; the calls coming from a single dashboard load. That gives the BE an excellent cache key!
+;;
+;; ### Why not cache on dashboard ID?
+;; There may be different users with different permissions fetching the same dashboard at the same time. They see a
+;; different picture of the queries and their metadata, so must be fetched separately.
+(def ^:private dashboard-load-cache-ttl
+  "Using 10 seconds for the cache TTL."
+  (* 10 1000))
+
+(def ^:private ^:dynamic *dashboard-load-id* nil)
+
+;; This is a kind of two-layer memoization:
+;; - The outer layer is a 10-second TTL cache on *dashboard-load-id*.
+;; - Its value is the *function* to use to get the dashboard by ID!
+;; If *dashboard-load-id* is set, the outer layer returns a forever-memoized wrapper around get-dashboard*.
+;; If *dashboard-load-id* is nil, it returns the unwrapped get-dashboard*.
+
+;; TODO: This indirect memoization by *dashboard-load-id* could probably be turned into a macro for reuse elsewhere.
```

Comment:
I suppose we could do this and then use this for embedded dashboards too (in api/embed)